### PR TITLE
ROS 2 dockerized example for nmea GPS

### PR DESCRIPTION
**Pull request by delihus**
Created ros2 workflow with standard date and package version tag.
Moved `compose.yaml` to `demo/` folder.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,24 +1,50 @@
-FROM ros:noetic-ros-core
+ARG ROS_DISTRO=humble
+ARG PREFIX=
 
+## =========================== ROS builder ===============================
```
Comment (pkowalsk1) [inline Dockerfile]: [major]
What's the purpose of having a separate entity building the package? The same packages are installed in both stages (i.e., ros-dev-tools), and rosdep is called twice with the same dependencies. In my opinion, this only complicates and unnecessarily prolongs the build process. I suggest installing and building everything sequentially in a single RUN command.

File: Dockerfile [hunk 2]
```diff
@@ -1,24 +1,50 @@
...
+
+WORKDIR /ros2_ws
+
+COPY --from=ros_builder /ros2_ws /ros2_ws
+COPY --from=ros_builder /version.txt /version.txt
+
+RUN apt-get update && apt-get install -y \
+        ros-dev-tools python3-transforms3d && \
+    rm -rf /etc/ros/rosdep/sources.list.d/20-default.list && \
+    rosdep init && \
```
Comment (pkowalsk1) [inline Dockerfile]: Is it really necessary? If certain dependencies are already included in the husarnet/ros image, I believe they should be installed directly. Moreover, could you elaborate on the advantages and disadvantages of using this particular image? What rationale led us to choose it for this relatively straightforward use case over the basic ros docker image?

File: Dockerfile [hunk 3]
```diff
@@ -1,24 +1,50 @@
...
+    rosdep update --rosdistro $ROS_DISTRO && \
+    rosdep install -i --from-path src --rosdistro $ROS_DISTRO -y && \
+    # Size optimalization
+    apt-get remove -y \
+        ros-dev-tools && \
     apt-get clean && \
+    rm -rf src && \
```
Comment (pkowalsk1) [inline Dockerfile]: I understand the importance of minimizing the image size, but in my opinion, this step may not be necessary. What do you think?

File: README.md [hunk 4]
```diff
@@ -1,20 +1,27 @@
-# nmea-gps-docker
-Building a Docker image for a GPS module
+<h1 align="center">
+  Docker Images for a GPS module
+</h1>
```
Comment (pkowalsk1) [inline README.md]:
```suggestion
# Docker Images for a GPS module
```
Alternatively, even the `# nmea-gps-docker`

File: README.md [hunk 5]
```diff
@@ -1,20 +1,27 @@
...
 
-### GPS API 
 
-GPS data in [NMEA](https://en.wikipedia.org/wiki/NMEA_0183) format is forwarded to RPi IP address at port 5000, typically it is `10.15.20.2:5000`. 
+The repository includes a GitHub Actions workflow that automatically deploys built Docker images to the [husarion/nmea-gps-docker](https://hub.docker.com/r/husarion/nmea-gps) Docker Hub repositories. This process is based on the [ros-drivers/nmea_navsat_driver](https://github.com/ros-drivers/nmea_navsat_driver/tree/ros2) repository.
+
+[![.github/workflows/build-docker-image.yaml](https://github.com/husarion/nmea-gps-docker/actions/workflows/build-docker-image.yaml/badge.svg?branch=ros2)](https://github.com/husarion/nmea-gps-docker/actions/workflows/build-docker-image.yaml)
+### GPS API
```
Comment (pkowalsk1) [inline README.md]: Can we align this ROS API and the whole README with our standards, like [here](https://hedgedoc.internal.husarion.com/readme-wrtiting) or [here]( https://github.com/husarion/panther_ros/tree/ros1/panther_driver#ros-nodes)?

File: .github/workflows/build-docker-image.yaml [hunk 6]
```diff
@@ -1,51 +1,64 @@
...
+      build_type:
+        description: Is it a "development" or a "stable" release?
         required: true
-        default: 'noetic'
-      branch:
-        description: 'branch that will be used to build image'
-        required: true
-        default: 'main'
-  
+        default: development
```
Comment (pkowalsk1) [inline .github/workflows/build-docker-image.yaml:10]: By the way, is it still in development or stable? I think we should aim to stabilize it and move away from reliance on this image. Are there plans to release tag v1.0.0 or something similar?

**Review by pkowalsk1 — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 1 above)
Comment (delihus) [reply Dockerfile]: There are some differences between `ros-core` and `ros-base` image but I can try to move this to one stage.

File: Dockerfile (hunk 2 above)
Comment (delihus) [reply Dockerfile]: Removing the sources is not necessary but rosdep has to be initialized other way it gives:
image

File: Dockerfile (hunk 3 above)
Comment (delihus) [reply Dockerfile]: I think it is good practice to reduce the image size as much as possible. In this case removing `src` does not impact a lot but e. g. when mesh files exists it changes a lot.

File: .github/workflows/build-docker-image.yaml (hunk 6 above)
Comment (delihus) [reply .github/workflows/build-docker-image.yaml:10]: In the docker repositories we added the development and the release build. Here are the options:
image

The purpose is to push images with date and package version.
image

After the PR is closed we can trigger the release build to make sure that image is tested and does work.
image
It adds to existing image e. g. `humble-2.0.1-20240209` `-stable` to the tag what gives `humble-2.0.1-20240209-stable`.

Referring to the version. The tag of images are versions of the built package described in `package.xml` and the date of docker image build.

Comment (pkowalsk1) [reply .github/workflows/build-docker-image.yaml:10]: I apologize for the misunderstanding. I've used this feature before, but I was referring to the GitHub tag.

Comment (delihus) [reply .github/workflows/build-docker-image.yaml:10]: okay! So after the PR I can release this!

**Review by pkowalsk1 — APPROVED** [review]

**Pull request merged**